docs/design/design_trivy_file_system_scanner.md

```diff
@@ -0,0 +1,200 @@
+  name: nginx
+  namespace: poc-ns
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      imagePullSecrets:
+        - name: private-registry
+      nodeName: kind-control-plane
+      containers:
+        - name: nginx
+          image: example.registry.com/nginx:1.16
+``` 
+
+ 
+***On Same Host***
+
+To scan the `nginx` container of such Deployment, Starboard could create the following scan job in the 
+`starboard-operator` namespace and observe it until it's completed or failed. Some points to note here
+- In this mode, we don't need imagePullSecret/any secret to pass to scan job
+- ImagePullPolicy should be set as Never
+- nodeName should be configured explicitly. 
+
+```yaml
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: scan-nginx-with-trivy
+  namespace: starboard-operator
+spec:
+  backoffLimit: 0
+  template:
+    spec:
+      restartPolicy: Never
+      # Explicit nodeName indicates our intention to schedule a scan pod
+      # on the same node where the nginx workload is running.
+      # This could also imply considering taints and tolerations and other
+      # properties respected by K8s scheduler.
+      nodeName: kind-control-plane
+      volumes:
+        - name: scan-volume
+          emptyDir: { }
+      initContainers:
+        # The trivy-get-binary init container is used to copy out the trivy executable
+        # binary from the upstream Trivy container image, i.e. aquasec/trivy:0.19.2,
+        # to a shared emptyDir volume.
+        - name: trivy-get-binary
+          image: aquasec/trivy:0.19.2
+          command:
+            - cp
+            - -v
+            - /usr/local/bin/trivy
+            - /var/starboard/trivy
+          volumeMounts:
+            - name: scan-volume
+              mountPath: /var/starboard
+         # The trivy-download-db container is using trivy executable binary
+         # from the previous step to download Trivy vulnerability database
+         # from GitHub releases page.
+        - name: trivy-download-db
+          image: aquasec/trivy:0.19.2
+          command:
+            - /var/starboard/trivy
+            - --download-db-only
+            - --cache-dir
+            - /var/starboard/trivy-db
+          volumeMounts:
+            - name: scan-volume
+              mountPath: /var/starboard
+      containers:
+        # The scan-nginx container is based on the container image that
+        # we want to scan with Trivy. However, it has overwritten command (entrypoint)
+        # to invoke trivy file system scan. The scan results are output to stdout
+        # in JSON format so we can parse them and store as VulnerabiltyReport instance.
+        - name: scan-nginx
+          image: example.registry.com/nginx:1.16
+          # To scan image layers cached on K8s node without pulling
+          # it from a remote registry.
+          imagePullPolicy: Never
+          command:
```

Comment:
We are changing the entrypoint here, but I think we should also mention that Trivy requires root permissions. IIRC a scan Job must run as UID 0. Does it have any impact on PSP or any solution that supersedes PSPs?

Reply:
Yes, we need to run this scan job container with root user `runAsUser: 0`. This will have impact with PSP MustRunAsNonRoot